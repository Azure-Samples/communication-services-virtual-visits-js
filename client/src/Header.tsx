// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import {
  Stack,
  Text,
  Panel,
  PanelType,
  Nav,
  INavLinkGroup,
  IconButton,
  ThemeContext,
  createTheme,
  IPanelProps
} from '@fluentui/react';
import { Theme } from '@fluentui/theme';
import React from 'react';
import {
  headerContainerStyles,
  headerMenuIconStyles,
  headerTextStyles,
  panelMenuIconStyles,
  wafflePanelNavigationStyles,
  wafflePanelStyles
} from './styles/Header.styles';

/**
 * @interface HeaderProps The properties that Header uses.
 * @member parentid The ID of the element that will be used to contain the panel. This is required so that the panel does not cover the entire height of the screen.
 */
interface HeaderProps {
  companyName: string;
  parentid: string;
}

interface WaffleMenuState {
  isOpen: boolean;
}

export class WaffleNavigation extends React.Component {
  render(): JSX.Element {
    const navLinkGroups: INavLinkGroup[] = [
      {
        links: [
          {
            name: 'Book',
            url: './book',
            key: 'book'
          },
          {
            name: 'Visit',
            url: './visit',
            key: 'visit'
          }
        ]
      }
    ];
    return <Nav ariaLabel="Navigate to expected page" styles={wafflePanelNavigationStyles} groups={navLinkGroups} />;
  }
}

export class WaffleMenu extends React.Component<{ parentid }, WaffleMenuState> {
  public constructor(props: { parentid }) {
    super(props);

    this.openPanel.bind(this);
    this.dismissPanel.bind(this);

    this.state = {
      isOpen: false
    };
  }

  public openPanel(): void {
    this.setState({ isOpen: true });
  }

  public dismissPanel(): void {
    this.setState({ isOpen: false });
  }

  private onRenderPanelMenu(props: IPanelProps | undefined, theme: Theme | undefined): JSX.Element {
    return (
      <>
        <IconButton
          id="waffle-menu-panel"
          iconProps={{ iconName: 'Waffle' }}
          aria-label="Menu Button"
          styles={theme ? panelMenuIconStyles(theme) : undefined}
          onClick={() => this.dismissPanel()}
        />
      </>
    );
  }

  render(): JSX.Element {
    return (
      <ThemeContext.Consumer>
        {(theme: Theme | undefined) => {
          if (theme === undefined) {
            theme = createTheme();
          }
          return (
            <div>
              <IconButton
                id="waffle-menu"
                iconProps={{ iconName: 'Waffle' }}
                aria-label="Menu Button"
                styles={headerMenuIconStyles(theme)}
                onClick={() => this.openPanel()}
              />
              <Panel
                styles={wafflePanelStyles(theme)}
                isLightDismiss={true}
                isOpen={this.state.isOpen}
                hasCloseButton={false}
                onDismiss={() => this.dismissPanel()}
                type={PanelType.smallFixedNear}
                layerProps={{ hostId: this.props.parentid }}
                onRenderNavigationContent={(props) => this.onRenderPanelMenu(props, theme)}
              >
                <WaffleNavigation />
              </Panel>
            </div>
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}

export class Header extends React.Component<HeaderProps> {
  public constructor(props: HeaderProps) {
    super(props);
  }

  render(): JSX.Element {
    return (
      <ThemeContext.Consumer>
        {(theme: Theme | undefined) => {
          if (theme === undefined) {
            theme = createTheme();
          }
          return (
            <Stack styles={headerContainerStyles(theme)} verticalAlign="center" horizontal>
              <WaffleMenu parentid={this.props.parentid} />
              <Text styles={headerTextStyles(theme)}>{this.props.companyName}</Text>
            </Stack>
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}
