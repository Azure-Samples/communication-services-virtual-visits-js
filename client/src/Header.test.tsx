// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { IconButton, Text, setIconOptions, Panel } from '@fluentui/react';
import { mount } from 'enzyme';
import { Header, WaffleNavigation } from './Header';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import * as renderer from 'react-test-renderer';

configure({ adapter: new Adapter() });

// Disable icon warnings for tests as we don't register the icons for unit tests which causes warnings.
// See: https://github.com/microsoft/fluentui/wiki/Using-icons#test-scenarios
setIconOptions({
  disableWarnings: true
});

describe('Header', () => {
  it('should render icon, waffle menu with panel, and company name', () => {
    const header = mount(<Header companyName="test" parentid="test" />);

    const waffleButton = header.find(IconButton);
    const companyText = header.find(Text);
    const waffleMenu = header.find('WaffleMenu');
    const panel = header.find(Panel);

    expect(waffleButton.length).toBe(1);
    expect(companyText.length).toBe(1);
    expect(waffleMenu.length).toBe(1);
    expect(panel.length).toBe(1);
  });

  it('panel should open when waffle icon is clicked on', () => {
    const header = mount(<Header companyName="test" parentid="test" />);

    const menuIcon = header.find({ iconName: 'Waffle' });
    const waffleMenu = header.find('WaffleMenu');
    waffleMenu.setState({ isOpen: false });

    menuIcon.simulate('click');
    expect(waffleMenu.state('isOpen')).toBe(true);
  });

  it('panel should have links to the book and visit page', () => {
    const waffleNavigationTree = renderer.create(<WaffleNavigation />).toJSON();
    expect(waffleNavigationTree).toMatchSnapshot();
  });
});
