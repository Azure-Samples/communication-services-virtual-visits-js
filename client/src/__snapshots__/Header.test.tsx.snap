// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header panel should have links to the book and visit page 1`] = `
<div
  className="ms-FocusZone css-151"
  data-focuszone-id="FocusZone8"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDownCapture={[Function]}
>
  <nav
    aria-label="Navigate to expected page"
    className="ms-Nav root-136"
    role="navigation"
  >
    <div
      className="ms-Nav-group is-expanded"
    >
      <div
        className="ms-Nav-groupContent groupContent-144"
      >
        <ul
          className="ms-Nav-navItems navItems-143"
          role="list"
        >
          <li
            className="ms-Nav-navItem navItem-142"
            role="listitem"
          >
            <div
              className="ms-Nav-compositeLink compositeLink-138"
              name="Book"
            >
              <a
                className="ms-Button ms-Button--action ms-Button--command ms-Nav-link link-152"
                data-is-focusable={true}
                href="./book"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                title="Book"
              >
                <span
                  className="ms-Button-flexContainer flexContainer-153"
                  data-automationid="splitbuttonprimary"
                >
                  <div
                    className="ms-Nav-linkText linkText-137"
                  >
                    Book
                  </div>
                </span>
              </a>
            </div>
          </li>
          <li
            className="ms-Nav-navItem navItem-142"
            role="listitem"
          >
            <div
              className="ms-Nav-compositeLink compositeLink-138"
              name="Visit"
            >
              <a
                className="ms-Button ms-Button--action ms-Button--command ms-Nav-link link-152"
                data-is-focusable={true}
                href="./visit"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                title="Visit"
              >
                <span
                  className="ms-Button-flexContainer flexContainer-153"
                  data-automationid="splitbuttonprimary"
                >
                  <div
                    className="ms-Nav-linkText linkText-137"
                  >
                    Visit
                  </div>
                </span>
              </a>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;
